rio_tiler_mosaic/mosaic.py

```diff
@@ -82,16 +100,25 @@ def mosaic_tiler(
             "'rio_tiler_mosaic.methods.base.MosaicMethodBase'"
         )
 
-    _tiler = partial(tiler, tile_x=tile_x, tile_y=tile_y, tile_z=tile_z, **kwargs)
-    max_threads = int(os.environ.get("MAX_THREADS", multiprocessing.cpu_count() * 5))
     if not chunk_size:
-        chunk_size = max_threads
+        chunk_size = threads or len(assets)
 
-    for chunks in _chunks(assets, chunk_size):
-        with futures.ThreadPoolExecutor(max_workers=max_threads) as executor:
-            future_tasks = [executor.submit(_tiler, asset) for asset in chunks]
+    tasks: Sequence[TaskType]
 
-        for t, m in _filter_futures(future_tasks):
+    for chunks in _chunks(assets, chunk_size):
+        if threads:
+            with futures.ThreadPoolExecutor(max_workers=threads) as executor:
+                tasks = [
+                    executor.submit(tiler, asset, tile_x, tile_y, tile_z, **kwargs)
+                    for asset in chunks
+                ]
+        else:
+            tasks = [
+                partial(tiler, asset, tile_x, tile_y, tile_z, **kwargs)
+                for asset in chunks
+            ]
```

Comment:
I think it might be a little cleaner as:
```py
tiler_fn = partial(tiler, tile_x=tile_x, tile_y=tile_y, tile_z=tile_z, **kwargs)

for chunks in _chunks(assets, chunk_size):
    if threads:
        with futures.ThreadPoolExecutor(max_workers=threads) as executor:
            tasks = [
                executor.submit(tiler_fn, asset)
                for asset in chunks
            ]
    else:
        tasks = (tiler_fn(asset) for asset in chunks)
```
that way it's clear that the same `tiler_fn` is being passed whether you're using threads or not. And I think by using a generator comprehension you still can stop early if you've finished your pixel selection method. Though you would need to change `yield future()` to `yield future` above I think.

Reply:
As discussed on slack, the problem with using keyword argument in partial is that you expect the `tiler` to have the `arguments` named `tile_x`, `tile_y`, `tile_z` which is a really hard requirement that I'd like to avoid. 

I think we can still change the list comprehension to generator ;-)

Reply:
well the Generator makes things way better, thanks for catching this @kylebarron